sw/device/silicon_creator/manuf/skus/earlgrey_a0/sival_bringup/ft_personalize.c

```diff
@@ -205,6 +215,81 @@ static void compute_keymgr_owner_binding(manuf_certgen_inputs_t *inputs) {
   memset(sealing_binding_value.data, 0, kDiceMeasurementSizeInBytes);
 }
 
+/**
+ *  A helper function to round up the passed in value to get it aligned to the
+ *  requested number of bits.
+ */
+static uint32_t round_up_to(uint32_t input, uint32_t align_bits) {
+  uint32_t mask = (1 << align_bits) - 1;
+
+  return (input + mask) & ~mask;
+}
+
+/**
+ *  Calculate the number of 4 byte words necessary to fit the passed in number
+ *  of bytes.
+ */
+static uint32_t size_to_words(uint32_t bytes) {
+  return round_up_to(bytes, 2) / sizeof(uint32_t);
+}
+
+/**
+ * Retrieve the certificate size from the passed in pointer to its ASN1 header.
+ * Perform some basic sanity checks.
+ */
+static uint32_t get_cert_size(const uint8_t *header) {
+  if (header[0] != 0x30 || header[1] != 0x82) {
+    return 0;
+  }
+
+  return (((uint32_t)header[2]) << 8) + header[3] + 4 /* size of the header */;
+}
+
+/**
+ *  Read a certificate from the passed in location in a flash INFO page and hash
+ *  its contents on the existing sha256 hashing stream. Determine the actual
+ *  certificate size from its ASN1 header.
+ *
+ * If the caller passed a pointer, save there the certificate size.
+ */
+static status_t hash_certificate(const flash_ctrl_info_page_t *page,
+                                 size_t offset, size_t *size) {
+  uint8_t buffer[1024];  // 1K should be enough for the largest possible
+                         // certificate.
+  uint32_t cert_size;
+  uint32_t bytes_read;  // Number of bytes already read.
+
+  // Read the first word of the certificate.
+  TRY(flash_ctrl_info_read(page, offset, 1, buffer));
+  bytes_read = sizeof(uint32_t);  // One word has been read.
+  cert_size = get_cert_size(buffer);
+  if (cert_size == 0) {
+    LOG_ERROR("Inconsistent certificate header %02x %02x page %x:%x", buffer[0],
+              buffer[1], page->base_addr, offset);
+    return DATA_LOSS();
+  }
+  if (cert_size > sizeof(buffer)) {
```

Comment:
Sorry one other thing: I think you want to use `ARRAYSIZE()` here: https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/base/macros.h;drc=30d7e787c753caaa03fe68a4a70da1bbcbc1d96f;l=45 no?

Reply:
never mind. pls disreged. cert_size is bytes b/c it is coming out of the ASN1 directly.